pep-0590.rst

```diff
@@ -48,145 +48,212 @@ The function pointer type
+So it makes a lot of sense to use vectorcall also for calling classes.
+This really means implementing the vectorcall protocol for ``type``.
+Some of the most commonly used classes will use this protocol,
+probably ``range``, ``list``, ``str``, and ``type``.
 
 The ``PyMethodDef`` protocol and Argument Clinic
-================================================
+------------------------------------------------
 
 Argument Clinic [4]_ automatically generates wrapper functions around lower-level callables, providing safe unboxing of primitive types and
 other safety checks.
-Argument Clinic could be extended to generate wrapper objects conforming to the new ``vectorcall`` protocol.
+Argument Clinic could be extended to generate wrapper objects with the ``METH_VECTORCALL`` signature.
 This will allow execution to flow from the caller to the Argument Clinic generated wrapper and
 thence to the hand-written code with only a single indirection.
 
+
+Third-party extension classes using vectorcall
+==============================================
+
+To enable call performance on a par with Python functions and built-in functions,
+third-party callables should include a ``vectorcallfunc`` function pointer,
+set ``tp_vectorcall_offset`` to the correct value and add the ``Py_TPFLAGS_HAVE_VECTORCALL`` flag.
+Any class that does this must implement the ``tp_call`` function and make sure its behaviour is consistent with the ``vectorcallfunc`` function.
+Setting ``tp_call`` to ``PyVectorcall_Call`` is sufficient.
+
+
 Performance implications of these changes
 =========================================
 
-Initial experiments, implementing the new calling convention for Python  functions, builtin functions and method-descriptors showed a
-speedup of around 2%. A full implementation involving other callables and adding support for the new calling convention to argument
-clinic would, in the author's estimation, yield a speedup of between 3% and 4% for the standard benchmark suite.
+This PEP should not have much impact on the performance of existing code
+(neither in the positive nor the negative sense).
+It is mainly meant to allow efficient new code to be written,
+not to make existing code faster.
+
+Nevertheless, this PEP optimizes for ``METH_FASTCALL`` functions.
+Performance of functions using ``METH_VARARGS`` will become slightly worse.
+
+
+Stable ABI
+==========
+
+Nothing from this PEP is added to the stable ABI (PEP 384).
 
 
 Alternative Suggestions
 =======================
 
+bpo-29259
+---------
+
+PEP 590 is close to what was proposed in bpo-29259 [#bpo29259]_.
+The main difference is that this PEP stores the function pointer
+in the instance rather than in the class.
+This makes more sense for implementing functions in C,
+where every instance corresponds to a different C function.
+
+The motivation is also different: this PEP is mainly meant to
+allow optimizating external classes (outside of CPython),
+not to optimize CPython itself.
```

Comment:
This PEP is meant to allow optimization of CPython as well.

Reply:
You mean specifically for `type` or do mean something else also?